### PR TITLE
Octave update

### PR DESCRIPTION
**Pull request by MarcusCalhoun-Lopez**
###### Type(s)

- [ ] bugfix
- [X] enhancement
- [ ] security fix

###### Tested on

macOS 10.13
Xcode 9.3

###### Verification 
Have you

- [X] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [X] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [X] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [X] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?

- [X] checked your Portfile with `port lint`?
- [X] tried existing tests with `sudo port test`?
- [X] tried a full install with `sudo port -vst install`?
- [X] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by MarcusCalhoun-Lopez** [comment]
In https://github.com/macports/macports-ports/pull/1827, @slewsys did all of the initial work on this issue (discussed on the [mailing list](https://lists.macports.org/pipermail/macports-dev/2018-May/038935.html)).
This PR diverges in a few ways:
- The patch files are a little different since they come from upstream
- GCC 7 is returned (attempts to fix [Qt issue](https://codereview.qt-project.org/#/c/221071/))
- Attempts to avoid [hardcoding version numbers](https://trac.macports.org/ticket/51707)
- Attempts to make sure all the Octave packages work
- Uses full path of makeinfo  instead of attempting to add to the path to find it
- Attempts to update dependencies
- Keeps octave-devel around
- FLTK is still around (this is an issue for debate)
- OpenMP is still disabled (I have been unable to test this issue yet)
- Probably a few others I have forgotten about

**Comment by ryandesign** [comment]
> Where did octave-spline-gcvspl come from?

See https://github.com/macports/macports-ports/commits/master/math/octave-spline-gcvspl/Portfile

**Comment by Schamschula** [comment]
In https://github.com/macports/macports-ports/commit/b80e0fa5a69e04e4ad2bf4a8842aacd69f54427b#diff-cf04665437847795836811d1f8cfce96 on Nov 27, 2013 it was noted that this package no longer exists on Octave Forge. We might want to mark it as obsolete.

**Comment by slewsys** [comment]
@MarcusCalhoun-Lopez Per [comment by Chih-Hsuan Yen](https://github.com/macports/macports-ports/pull/1827#issuecomment-389738596), until Travis timeout period is extended, tell Travis not to even attempt compiling Octave by adding `[ci skip]` or `[skip ci]` to commit message.

Thank you for identifying GCC and INIT_PRIORITY issue.  Might it be worth moving patch to a separate PR, since it's really a Qt issue (and referencing this PR in commit message)?

Until I have a chance to build this PR, I'm wondering:

* My recollection is that port `hdf5` has to be built with variant `openmpi` disabled in order for Octave configure to find it, but that Octave itself builds fine with multi-threading enabled (i.e., configure option `--enable-openmp`).  Am I wrong, or is there a particular reason for using `--disable-openmp`?

* Octave 4.4 needs a newer version of `makeinfo(1)` than provided by /usr/bin/makeinfo for Octave's `help` command to work.  This isn't an issue if octave is run from the command line, since it inherits the login shell's PATH.  However without an initialization file, if Octave is launched in the Finder by double-clicking, e.g.,  _/Applications/MacPorts/Octave.app_, then it inherits launchd's PATH, as seen by the output of the Octave command `getenv ("PATH")`.

* With this PR, does  Octave command `graphics_toolkit('fltk')` not cause Octave to crash?  If not, which version of FLTK works with Octave?

In the mean time, thank you for all the updates!

**Comment by pmetzger** [comment]
Should we close the earlier PR now?

**Comment by slewsys** [comment]
On May 25, 2018, at 9:10 AM, Perry E. Metzger <notifications@github.com> wrote:
> 
> Should we close the earlier PR now?
> 
Yes

**Comment by pmetzger** [comment]
@slewsys Please do then. You can just bring it up and "close and comment" at the bottom.

**Comment by MarcusCalhoun-Lopez** [comment]
> My recollection is that port hdf5 has to be built with variant openmpi disabled in order for Octave configure to find it, but that Octave itself builds fine with multi-threading enabled (i.e., configure option --enable-openmp).

When hdf5 is built with an MPI variant, it is not found by octave because the configure test fails with
`fatal error: 'mpi.h' file not found` error.
Getting this to work would probably require a fair bit of testing.
For example, it is possible that octave would have to be built with the **same** MPI variant as hdf5.
I am not sure how `--enable-openmp` would have any affect.
[MPI](https://en.wikipedia.org/wiki/Message_Passing_Interface) and [OpenMP](https://en.wikipedia.org/wiki/OpenMP) are very different beasts.

> Am I wrong, or is there a particular reason for using --disable-openmp?

For the time being, I left `--disable-openmp` just because it has not been very well tested yet.
Plus, as has been noted, it would only affect legacy OSs and those who specifically request a different compiler.
Please correct me if I am wrong.

> Octave 4.4 needs a newer version of makeinfo(1) than provided by /usr/bin/makeinfo for Octave's help command to work.

Instead of modifying the `PATH` variable, I had hoped to take care of this by patching `libinterp/corefcn/help.h`. The `patch-makeinfo.diff` file uses the full pathname of `makeinfo`.

> With this PR, does Octave command graphics_toolkit('fltk') not cause Octave to crash?

FLTK still causes a crash unless `--no-gui-libs` is passed to octave.
As has been [noted](https://github.com/macports/macports-ports/pull/1827#issuecomment-390414710), FLTK may be on the way out, but in the same [discussion](http://octave.1599824.n4.nabble.com/Deprecate-fltk-graphics-toolkit-tp4685506p4685535.html), there is some support of a "minimal version of Octave."
For our purposes, perhaps gnuplot satisfies that requirement.

**Comment by MarcusCalhoun-Lopez** [comment]
> I haven't found any way of building octave against hdf5 with openmpi variant. I'd be tempted to add a comment to that effect in the Portfile.

There is now [a comment](https://github.com/macports/macports-ports/pull/1865/commits/1c6bf8bc39fe24ece1bff675fdb15d007ab676af) and the Portfile checks the hdf5 variant.

>Looking into this further, I see that the Matlab/Octave API isn't thread safe, so disabling option openmp by default seems reasonable.

Thanks for looking into this.

> I'd be concerned that, without addressing the issue, including FLTK by default might give the impression that Octave GUI is not functional when the test suite crashes.

Fair point.
FLTK is [no longer included by default](https://github.com/macports/macports-ports/pull/1865/commits/9e7d9db47b00a68cb7e84fc0d987eb4bceff0156).
I had to rename the variant or else every upgrading would automatically get it.

**Comment by MarcusCalhoun-Lopez** [comment]
I have not hear any concerns from the Octave package maintainers.
Have all the major issues been addressed?
If so, we can go ahead and merge this pull request.

**Comment by pmetzger** [comment]
@MarcusCalhoun-Lopez I think they have, but I'd let it run a full 24 hours to hear contrary opinions, then merge.

File: math/octave/Portfile [hunk 1]
```diff
@@ -51,15 +53,18 @@ patchfiles-append patch-link.diff
...
-patchfiles-append patch-host_name.diff
+patchfiles-append patch-versions.diff
 
-# see https://savannah.gnu.org/patch/index.php?9439
-patchfiles-append patch-java.diff
+# reports of `__opengl_info__` not working the first time
+patchfiles-append patch-__opengl_info__.diff
 
 post-patch {
     if { ${os.arch} eq "i386" } {
```
Comment (slewsys) [inline math/octave/Portfile]: Don't know if it even matters, but I updated this line to:
```
    if { ${os.arch} eq "i386" || ${os.arch} eq "x86_64" } {
```

Comment (MarcusCalhoun-Lopez) [reply math/octave/Portfile]: I believe ``${os.arch}`` can only be ``powerpc`` or ``i386`` (see [here](https://github.com/macports/macports-base/blob/556a929a3917bd0a2c02166e916c2e4cc5f99172/src/macports1.0/macports.tcl#L673)).

**Comment by slewsys** [comment]
On Jun 1, 2018, at 10:28 PM, MarcusCalhoun-Lopez <notifications@github.com> wrote:
> 
> @MarcusCalhoun-Lopez commented on this pull request.
> 
> In math/octave/Portfile:
> 
> >  
> -# see https://savannah.gnu.org/patch/index.php?9439
> -patchfiles-append patch-java.diff
> +# reports of `__opengl_info__` not working the first time
> +patchfiles-append patch-__opengl_info__.diff
>  
>  post-patch {
>      if { ${os.arch} eq "i386" } {
> 
> I believe ${os.arch} can only be powerpc or i386 (see here).

You are right.  Should have confirmed before posting.  My apologies.
-AM

**Comment by pmetzger** [comment]
@MarcusCalhoun-Lopez Probably time to merge?

**Pull request merged**